testsuite/tests/typing-modes/statefulness-visibility.ml

```diff
@@ -0,0 +1,552 @@
+Line 1, characters 55-56:
+1 | let fails : 'a @ observing -> 'a @ portable = fun x -> x
+                                                           ^
+Error: This value is "nonportable" but expected to be "portable".
+|}]
+
+let succeeds : 'a @ observing portable -> 'a @ portable = fun x -> x
+[%%expect{|
+val succeeds : 'a @ portable observing -> 'a @ portable = <fun>
+|}]
+
+let fails : 'a @ stateful -> 'a @ portable = fun x -> x
+[%%expect{|
+Line 1, characters 54-55:
+1 | let fails : 'a @ stateful -> 'a @ portable = fun x -> x
+                                                          ^
+Error: This value is "nonportable" but expected to be "portable".
+|}]
+
+let succeeds : 'a @ stateful portable -> 'a @ portable = fun x -> x
+[%%expect{|
+val succeeds : 'a @ portable -> 'a @ portable = <fun>
+|}]
+
+(* Modalities. *)
+
+type 'a t = { x : 'a @@ stateless }
+
+let get : 'a t -> 'a @ portable = fun t -> t.x
+
+[%%expect{|
+type 'a t = { x : 'a @@ stateless; }
+val get : 'a t -> 'a @ portable = <fun>
+|}]
+
+(* [immutable] => [contended]. *)
+
+let default : 'a @ contended -> ('a @ immutable -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+val default : 'a @ contended -> ('a @ immutable -> 'b) -> 'b = <fun>
+|}]
+
+let override : 'a @ contended -> ('a @ immutable uncontended -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 89-90:
+1 | let override : 'a @ contended -> ('a @ immutable uncontended -> 'b) -> 'b = fun x f -> f x
+                                                                                             ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+let override : 'a @ contended -> ('a @ immutable shared -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 84-85:
+1 | let override : 'a @ contended -> ('a @ immutable shared -> 'b) -> 'b = fun x f -> f x
+                                                                                        ^
+Error: This value is "contended" but expected to be "shared".
+|}]
+
+(* [read] => [shared]. *)
+
+let default : 'a @ contended -> ('a @ read -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 71-72:
+1 | let default : 'a @ contended -> ('a @ read -> 'b) -> 'b = fun x f -> f x
+                                                                           ^
+Error: This value is "contended" but expected to be "shared".
+|}]
+
+let override : 'a @ contended -> ('a @ read uncontended -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+Line 1, characters 84-85:
+1 | let override : 'a @ contended -> ('a @ read uncontended -> 'b) -> 'b = fun x f -> f x
+                                                                                        ^
+Error: This value is "contended" but expected to be "uncontended".
+|}]
+
+let override : 'a @ contended -> ('a @ read contended -> 'b) -> 'b = fun x f -> f x
+[%%expect{|
+val override : 'a @ contended -> ('a @ contended read -> 'b) -> 'b = <fun>
+|}]
+
+(* [read_write] doesn't change the default. *)
```

Comment:
This should probably include tests showing that `read_write` on its own can't be used as `contended` or `shared`

Reply:
Done.